types/eslint/index.d.ts

```diff
@@ -241,7 +241,10 @@ export namespace Rule {
         meta?: RuleMetaData;
     }
 
-    interface RuleListener {
+    type NodeTypes = ESTree.Node['type'];
+    type NodeListener = { [T in NodeTypes]?: (node: ESTree.Node) => void };
```

Comment:
Would it be possible to type `node` to be `T`?

(I'm a TS newbie, so might not be needed).

```ts
const rule: Rule.RuleModule = {
  create() {
    return {
      CallExpression(node: CallExpression) {
        console.log(node.callee);
      },
    };
  },
};
```
Gives me `Type 'Node' is not assignable to type 'CallExpression'`

Reply:
The solution to this is a two-parter:

1. <details><summary>Add a new mapping to <code>@types/estree</code>. This maps from the name of the node to its type.</summary>

   ```ts
   // yes, this is necessary :(
   type _NodeMap = { [K in Node['type']]: Node }

   export interface NodeByType extends _NodeMap {
     Identifier: Identifier

     SimpleLiteral: SimpleLiteral
     RegExpLiteral: RegExpLiteral

     Program: Program

     FunctionDeclaration: FunctionDeclaration
     FunctionExpression: FunctionExpression
     ArrowFunctionExpression: ArrowFunctionExpression

     SwitchCase: SwitchCase
     CatchClause: CatchClause
     VariableDeclarator: VariableDeclarator

     ExpressionStatement: ExpressionStatement
     BlockStatement: BlockStatement
     EmptyStatement: EmptyStatement
     DebuggerStatement: DebuggerStatement
     WithStatement: WithStatement
     ReturnStatement: ReturnStatement
     LabeledStatement: LabeledStatement
     BreakStatement: BreakStatement
     ContinueStatement: ContinueStatement
     IfStatement: IfStatement
     SwitchStatement: SwitchStatement
     ThrowStatement: ThrowStatement
     TryStatement: TryStatement
     WhileStatement: WhileStatement
     DoWhileStatement: DoWhileStatement
     ForStatement: ForStatement
     ForInStatement: ForInStatement
     ForOfStatement: ForOfStatement

     // FunctionDeclaration: FunctionDeclaration
     VariableDeclaration: VariableDeclaration
     ClassDeclaration: ClassDeclaration

     ThisExpression: ThisExpression
     ArrayExpression: ArrayExpression
     ObjectExpression: ObjectExpression
     // FunctionExpression: FunctionExpression
     // ArrowFunctionExpression: ArrowFunctionExpression
     YieldExpression: YieldExpression
     // Literal: Literal
     UnaryExpression: UnaryExpression
     UpdateExpression: UpdateExpression
     BinaryExpression: BinaryExpression
     AssignmentExpression: AssignmentExpression
     LogicalExpression: LogicalExpression
     MemberExpression: MemberExpression
     ConditionalExpression: ConditionalExpression
     CallExpression: CallExpression
     NewExpression: NewExpression
     SequenceExpression: SequenceExpression
     TemplateLiteral: TemplateLiteral
     TaggedTemplateExpression: TaggedTemplateExpression
     ClassExpression: ClassExpression
     MetaProperty: MetaProperty
     // Identifier: Identifier
     AwaitExpression: AwaitExpression

     Property: Property
     AssignmentProperty: AssignmentProperty
     Super: Super
     TemplateElement: TemplateElement
     SpreadElement: SpreadElement

     // Identifier: Identifier
     ObjectPattern: ObjectPattern
     ArrayPattern: ArrayPattern
     RestElement: RestElement
     AssignmentPattern: AssignmentPattern
     // MemberExpression: MemberExpression

     ClassBody: ClassBody
     // ClassDeclaration: ClassDeclaration
     // ClassExpression: ClassExpression
     MethodDefinition: MethodDefinition

     ImportDeclaration: ImportDeclaration
     ExportNamedDeclaration: ExportNamedDeclaration
     ExportDefaultDeclaration: ExportDefaultDeclaration
     ExportAllDeclaration: ExportAllDeclaration

     ImportSpecifier: ImportSpecifier
     ImportDefaultSpecifier: ImportDefaultSpecifier
     ImportNamespaceSpecifier: ImportNamespaceSpecifier
     ExportSpecifier: ExportSpecifier
   }
   ```

   </details>

1. Redefine the listener:

   ```ts
   type NodeListener = {
     [T in ESTree.Node['type']]?: (node: ESTree.NodeByType[T]) => void
   }
   ```


I defined the mapping in `@types/estree` so it gets updated as new nodes are added.

Reply:
@j-f1 could you send a PR for that? 👼

Reply:
@SimenB I got started [here](https://github.com/DefinitelyTyped/DefinitelyTyped/commit/a9ac72f3ebfe6adbcfffd728a092972b8bb3a7d2), but I couldn’t get the tests to pass :(